hydra-node/test/Hydra/Chain/Direct/TxSpec.hs

```diff
@@ -141,6 +187,215 @@ spec =
+                      commitOutputs = toList $ commitTxBody ^. outputsTxBodyL
+                   in property
+                        ( all
+                            (`notElem` blueprintOutputs)
+                            commitOutputs
+                        )
+                        & counterexample ("blueprint outputs: " <> show blueprintOutputs)
+                        & counterexample ("commit outputs: " <> show commitOutputs)
+                , let blueprintSigs = blueprintBody ^. reqSignerHashesTxBodyL
+                      commitSigs = commitTxBody ^. reqSignerHashesTxBodyL
+                   in property (blueprintSigs `Set.isSubsetOf` commitSigs)
+                        & counterexample ("blueprint signatures: " <> show blueprintSigs)
+                        & counterexample ("commit signatures: " <> show commitSigs)
+                , let blueprintRefInputs = blueprintBody ^. referenceInputsTxBodyL
+                      commitRefInputs = commitTxBody ^. referenceInputsTxBodyL
+                   in property (blueprintRefInputs `Set.isSubsetOf` commitRefInputs)
+                        & counterexample ("blueprint reference inputs: " <> show blueprintRefInputs)
+                        & counterexample ("commit reference inputs: " <> show commitRefInputs)
+                ]
+                & cover 1 (spendsFromScript (lookupUTxO, fromLedgerTx blueprintTx)) "blueprint spends script UTxO"
+                & cover 1 (spendsFromPubKey (lookupUTxO, fromLedgerTx blueprintTx)) "blueprint spends pub key UTxO"
+                & cover 1 (not . null $ blueprintTx ^. bodyTxL . referenceInputsTxBodyL) "blueprint has reference input"
+
+getAuxMetadata :: AlonzoTxAuxData LedgerEra -> Map Word64 Metadatum
+getAuxMetadata (AlonzoTxAuxData metadata _ _) = metadata
+
+addRequiredSignatures :: [VerificationKey PaymentKey] -> Tx -> Tx
+addRequiredSignatures vks tx =
+  fromLedgerTx $
+    toLedgerTx tx
+      & bodyTxL . reqSignerHashesTxBodyL .~ Set.fromList (toLedgerKeyHash . verificationKeyHash <$> vks)
+
+genBlueprintTxWithUTxO :: Gen (UTxO, Tx)
+genBlueprintTxWithUTxO =
+  fmap (second unsafeBuildTransaction) $
+    spendingPubKeyOutput (mempty, emptyTxBody)
+      >>= spendSomeScriptInputs
+      >>= addSomeReferenceInputs
+      >>= addValidityRange
+      >>= addRandomMetadata
+      >>= removeRandomInputs
+      >>= addCollateralInputs
+ where
+  spendingPubKeyOutput (utxo, txbody) = do
+    utxoToSpend <- (genUTxOAdaOnlyOfSize . getPositive) . Positive =<< choose (0, 50)
+    pure
+      ( utxo <> utxoToSpend
+      , txbody & addVkInputs (toList $ UTxO.inputSet utxoToSpend)
+      )
+
+  spendSomeScriptInputs (utxo, txbody) = do
+    let alwaysSucceedingScript = PlutusScriptSerialised $ Plutus.alwaysSucceedingNAryFunction 3
+    n <- choose (1, 100)
+    datum <- unsafeHashableScriptData . fromPlutusData . I <$> arbitrary
+    redeemer <-
+      unsafeHashableScriptData . fromPlutusData . B . BS.pack
+        <$> vector n
```

Comment:
Could: use more arbitrary datum and redeemers

Did you decide for only integers and only bytestrings for a reason? Maybe `arbitrary` datums/redeemers are fine too?

For example, as a reader it's not clear why the length of the redeemer bytestring would be equal to the number of script inputs we generate.

In any case, leaving a comment why the single "always succeeding script" and one type of datum / redeemer is okay would be great.